### ISSUE TITLE
Handling long clock jumps in PBTS

### ISSUE THREAD

**Issue by lasarojc** [cometbft/cometbft#2521]
This issue is for discussing the effects of long clock jumps, backward and forward, in a chain using PBTS.

PBTS assumes nodes are synchronized to a source of truth clock using NTP. Most likely the source of truth is UTC and synchronization is achieved through the many NTP servers public accessible. 
If an attacker were to subvert the set of NTP servers used by the validators in a chain, what would be the effects?

The following is first flush of ideas and contributions are welcome.

## DoS attacks

A DoS attack would likely have a lesser impact. NTP can be configured with multiple servers to fallback, but even if that's not the case, nodes would likely remain in synch for long periods, making the DoS attack expensive.

## Backwards jumps

An attack that moves the NTP clock backwards could impact the chains, if backwards time jumps are allowed in the validator nodes. This happens because block times need to be monotonically increasing and PBTS ensures this by forcing proposers in a PBTS enabled chain to wait until their clocks surpasses the timestamp of the latest block decided before proposing. Hence, in this scenario, the longer the backwards jump, the longer the chain halt.

Fixing this problem requires only jumping the clock ahead once again, so the catchup happens quickly. Once this happens and local clocks are adjusted by the next proposer, it will quickly propose a block. 

In the wake of the forward jump, proposals may be rejected until a quorum of validators fixes their clocks and/or the adaptiveness of PBTS counters the clock desynchronization.

## Forward jumps

Forward jumps do not halt the network, but causes block times to also jump ahead.
Fixing the clocks, however, is more complicated.
If clocks smearing is used to adjust the time backwards, the network will keep making progress, but block times will continue not to match the real time/UTC for the duration of the smearing.
If the clock is quickly jumped backwards, then again the proposers will not propose until the clocks catch up with the latest block timestamps. If neither of these approaches is palatable, then a third option is to hard fork the chain to reset the block times.

**Comment by andynog** [comment]
In the case of Backward jump, is it possible to reject the block proposal if the timestamp of the proposed block is older than the last block committed ?

**Comment by cason** [comment]
> is it possible to reject the block proposal if the timestamp of the proposed block is older than the last block committed ?

Yes, this is a [safety check ](https://github.com/cometbft/cometbft/blob/main/internal/state/validation.go#L116) performed during the validation of a block. It was present in BFT Time and it wasn't changed or disabled with the introduction of PBTS.

**Comment by odeke-em** [comment]
I have created a package https://github.com/orijtech/rtset which I can gladly donate to the Cosmos community (I just needed to stand it up at some location) but importing it and using it is super trivial (you'll need sudo su permissions though) to periodically synchronize your time with the Roughtime server and that solves the problem for every validator needing to have a secure way to synchronize their time, here is an example you can run locally on your computer
```go
```

**Comment by sergio-mena** [comment]
@odeke-em I took a quick peek at your repo. I understand it leverages Cloudflare's "Roughtime" service.

After checking Cloudflare's documentation, I think it has a pro and two cons w.r.t. well-known NTP:

* Pro: looks simpler to secure as it uses modern crypto out of the box
* Con1: it is _rough_ rather than precise. The documentation says _within 10 seconds of real time_. This might be problematic for certain PBTS configurations, as 10 secs is too much
* Con2: We depend on Cloudflare as a service to be up (unless there are similar services offered out there, didn't check myself)

Additionally, on the "Pro" above, even if it is usually harder to access NTP servers securely, a sensible network operator would have _many_ NTP servers to make a "fake time" attack much harder.

**Comment by adizere** [comment]
I think we should not enshrine a Cloudfare-based for time synchronization, but it's a great option to have for app developers who feel confident using it, nice work @odeke-em !

**Comment by cason** [comment]
Probably an action item for this issue is to describe the problem in the documentation.

We are introducing clock synchronization as a liveness requirement. The clock synchronization should therefore be fault tolerant. This means, in particular, that failures in the clock synchronization mechanism must be independent. This can be achieved by relying on distinct clock synchronization services/servers or by adding, as a secondary component, a solution like "Roughtime".

As @sergio-mena mentioned above, this protocol does not provide good granularity and/or precision. So it probably should be combined with ordinary NTP protocols, with the former being used to actually synchronize the validators' clocks, while the latter being used to periodically validate the "reasonableness" of the validator's local clock.

The goal is therefore not to use services such as "Roughtime" as a source of time, but as a way to detect attacks in the adopted time service.

**Comment by cason** [comment]
> The goal is therefore not to use services such as "Roughtime" as a source of time, but as a way to detect attacks in the adopted time service.

By the way, we don't really need the validators to run this service, it could be ran by clients or external service, in order to validate the produced block times.

**Comment by sambacha** [comment]
Hello, I would like to mention some concerns regarding the usage of 'Roughtime'

> I think we should not enshrine a Cloudfare-based for time synchronization, but it's a great option to have for app developers who feel confident using it, nice work [@odeke-em](https://github.com/odeke-em) !

I have to agree with this prudent statement: Cloudflare already has a large network footprint, i.e. https://bgp.tools/as/13335 has a large cross-section of nodes. It may be prudent to not further provide networking overlap to Cloudflare.


Roughtime does not have the operational success when compared to NTP vis a vie ntp protocol operators. 

Here is an expert from an incident that occured directly because of the usage of Roughtime:

> Notice something off about this? The 6th server in the list is reporting a time which is 24 hours further than the other 5 servers. When the roughtime client took the average of these results, it thought that the correct time was Fri, 14 Aug 2020 22:20:23 GMT. One of the key components of the roughtime protocol is accountability. Given that there is a signed “clockchain” of sorts, we could prove that the ticktock server misbehaved and reported the issue effectively. Unfortunately, Prysm did not log the signed chain and we do not have it stored anywhere from the time of the event. After a quick look at the logs, TickTock was off by 24 hours which would explain why we had a 4 hour offset with 6 servers. We got in contact with the maintainers of the roughtime project from cloudflare and they have since determined how to make their client more robust in case of server faults[^1]


With the Marzullo algorithm implementation, NTPv4 can tolerate up to `⌊(m-1)/3⌋` faulty servers where `m` represents the total server count[^2]. The implementation generates a complete statistical model of candidate clocks, identifying the largest clique of mutually agreeing sources while systematically excluding outliers (i.e. its BFT-ish?).

For what it's worth specifying what supported systems you would want would help as not all OS's support, e.g:

- System sleep can happen in the middle of executing that code.
- System sleep affects how *tick* can operate differently on separate platforms:
Windows includes the system sleep time in the passage of time.
Linux/Android do not.




[^1]: https://medium.com/prysmatic-labs/eth2-medalla-testnet-incident-f7fbc3cc934a
[^2]: https://www.ietf.org/rfc/rfc5905.txt
